pkg/codegen/dotnet/gen.go

```diff
@@ -1021,6 +1021,72 @@ func (mod *modContext) genEnums(w io.Writer, enums []*schema.EnumType) error {
 	return nil
 }
 
+// genResourcePackage generates a ResourcePackage definition. The generated ResourcePackage supports
+// the deserialization of resource references into fully-hydrated Resource instances and providers.
+func genResourcePackage(pkg *schema.Package, assemblyName string, modules map[string]*modContext) []byte {
+	contract.Assert(len(modules) != 0)
+
+	w := &bytes.Buffer{}
+
+	// Generate a header and find the provider.
+	var provider *schema.Resource
+	hasHeader := false
+	for _, mod := range modules {
+		if !hasHeader {
+			mod.genHeader(w, []string{"System"})
+			hasHeader = true
+		}
+		for _, r := range mod.resources {
+			if r.IsProvider {
+				contract.Assert(provider == nil)
+				provider = r
+				continue
+			}
+		}
+	}
+
+	// Open the namespace.
+	fmt.Fprintf(w, "namespace %s\n", assemblyName)
+	fmt.Fprintf(w, "{\n")
+
+	fmt.Fprintf(w, "    internal class ResourcePackage : IResourcePackage\n")
+	fmt.Fprintf(w, "    {\n")
+	fmt.Fprintf(w, "        public string Name => \"%s\";\n\n", pkg.Name)
+	fmt.Fprintf(w, "        public string? Version => null;\n\n")
```

Comment:
I added version info to Node.js and Python in https://github.com/pulumi/pulumi/pull/5839. Want to do the same here?